README.md
# GAN based data augmentation approach of multi-featured and multi-dimensional time-series wearable motion data for robust HMR with DCNN

In this paper, GAN-based data augmentation is the subject of study to use multi-feathered and multi-dimensional time-series wearable motion data for DCNN. 

For the study, vanilla GAN implemented by Pytorch open source library was used, and loss graph and output were compared to be analyzed.

Research has shown how the application of GAN works on data converted from sign to image. The results showed that there was not enough quality to be used, even if it had some effect.

However, assuming that more advanced form of DCGAN or yLSGAN is used, the output is expected to be available as the actual training data.

## 1. 서론

### 1.1 연구의 배경 및 필요성
HMR(Human Motion Recognition)은 변위센서, 속도센서, 토크센서 등 모션 인식 센서로부터 얻은 데이터로부터 대상이 어떠한 동작을 취하고 있는지 알아내는 것을 말한다. HMR을 하기위한 방법에는 여러 가지가 있지만, 그 중에서 DCNN(Deep Convolution Neural Network)을 적용해 결과를 낸다고 가정한다.

DCNN(Deep Convolution Neural Network)은 합성곱(Convolution)연산을 사용하는 인공 신경망의 한 종류이며, CNN에서 neural network layer의 수를 늘려 더 깊어진 형태의 인공 신경망을 의미한다.

DCNN의 model을 학습시키기 위해서 행렬 형태의 이미지가 train data로써 입력된다. 이 때, 훈련할 충분한 학습 데이터를 확보하지 않으면, model의 성능을 저해하는 overfitting 문제가 발생할 수 있다. 따라서 정확한 DCNN model을 만들기 위해서는 양질의 데이터를 대량으로 확보할 수 있어야 한다. 학습 데이터를 만들기 위해 기본적으로 실험 데이터를 얻어와 전처리(preprocessing) 하지만, 비용적인 문제 때문에 실험만으로 모든 데이터를 얻는 데에는 무리가 있다. 대상으로 삼을 시계열 time-series weable motion data 역시, 실험 시간이 오래 걸리는 점, 매 실험마다 데이터에 통일성을 부여하기 위해 전처리를 따로 해야한다는 점 등등 시간과 비용적인 문제와 직면한다.
#### 1.1.1 연구의 배경
#### 1.1.2 연구의 필요성
#### 1.2 연구의 목표 및 방법

## 2. 배경 이론
### 2.1 GAN(Generative Adversarial Network)
#### 2.1.1 GAN의 개념
#### 2.1.2 GAN의 학습과정
#### 2.1.3 GAN의 loss function
### 2.2 Input Parameters
#### 2.2.1 Learning Rate
#### 2.2.2 Batch Size
#### 2.2.3 Epoch
### 2.3 Training Dataset

## 3. Data augmentation
### 3.1 Case01 : Initial Condition
### 3.2 Case02 : Changed Learning rate of G & D
### 3.3 Case03 : Changed Learning rate of G
### 3.4 Case04 : Changed Batch Size
### 3.5 Final Case : Changed Epochs
#### 3.5.1 구간 분석
#### 3.5.2 Training data & Output 비교
##### 3.5.2.1 Training data
##### 3.5.2.2 Output

## 4. 결 론


history.txt
batch_size = 512	learning_rate_g = 0.00005	learning_rate_d = 0.00005	epochs = 300
batch_size = 512	learning_rate_g = 0.0001	learning_rate_d = 0.0001	epochs = 300
batch_size = 512	learning_rate_g = 0.0002	learning_rate_d = 0.0002	epochs = 300
batch_size = 512	learning_rate_g = 0.0003	learning_rate_d = 0.0003	epochs = 300
batch_size = 512	learning_rate_g = 0.0004	learning_rate_d = 0.0004	epochs = 300
batch_size = 512	learning_rate_g = 0.0005	learning_rate_d = 0.0005	epochs = 300

----------------------------------------------learning rate를 다르게 주면서 해보았다. 0.0002에서 0.0003이 적당하다고 판단

batch_size = 512	learning_rate_g = 0.0001	learning_rate_d = 0.0003	epochs = 300
batch_size = 512	learning_rate_g = 0.0003	learning_rate_d = 0.0001	epochs = 300
batch_size = 512	learning_rate_g = 0.0003	learning_rate_d = 0.0002	epochs = 300
batch_size = 512	learning_rate_g = 0.0003	learning_rate_d = 0.0003	epochs = 600

batch_size = 512	learning_rate_g = 0.0002	learning_rate_d = 0.0002	epochs = 600
batch_size = 512	learning_rate_g = 0.0002	learning_rate_d = 0.0002	epochs = 900

batch_size = 1024	learning_rate_g = 0.0002	learning_rate_d = 0.0002	epochs = 600
batch_size = 1024	learning_rate_g = 0.0003	learning_rate_d = 0.0002	epochs = 600

batch_size = 1024	learning_rate_g = 0.0003	learning_rate_d = 0.0002	epochs = 1200
batch_size = 1024	learning_rate_g = 0.0003	learning_rate_d = 0.0002	epochs = 2400



batch_size = 256	learning_rate_g = 0.0001	learning_rate_d = 0.00005	epochs = 1200

batch_size = 256	learning_rate_g = 0.0001	learning_rate_d = 0.00005	epochs = 3600




------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------------------------








gan.py
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torch.optim as optim
import torchvision.datasets as datasets
import imageio
import numpy as np
import matplotlib

from torchvision.utils import make_grid, save_image
from torch.utils.data import DataLoader
from matplotlib import pyplot as plt
from tqdm import tqdm

matplotlib.style.use('ggplot')

# learning parameters
batch_size = 512
learning_rate_g = 0.00005
learning_rate_d = 0.00005
epochs = 300


sample_size = 64 # fixed sample size
nz = 128 # latent vector size
k = 1 # number of steps to apply to the discriminator
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

transform = transforms.Compose([
                                transforms.ToTensor(),
                                transforms.Normalize((0.5,),(0.5,)),
])

to_pil_image = transforms.ToPILImage()

train_data = datasets.ImageFolder(
    root='input/data/motion',
    transform=transform
)
train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True)

class Generator(nn.Module):
    def __init__(self, nz):
        super(Generator, self).__init__()
        self.nz = nz
        self.main = nn.Sequential(
            nn.Linear(self.nz, 256),
            nn.LeakyReLU(0.2),

            nn.Linear(256, 512),
            nn.LeakyReLU(0.2),

            nn.Linear(512, 1024),
            nn.LeakyReLU(0.2),

            nn.Linear(1024, 924),
            nn.Tanh(),
        )

    def forward(self, x):
        return self.main(x).view(-1, 3, 22, 14)

class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()
        self.n_input = 924
        self.main = nn.Sequential(
            nn.Linear(self.n_input, 1024),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),

            nn.Linear(1024, 512),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),

            nn.Linear(512, 256),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),

            nn.Linear(256, 1),
            nn.Sigmoid(),
        )

    def forward(self, x):
        x = x.view(-1, 924)
        return self.main(x)

generator = Generator(nz).to(device)
discriminator = Discriminator().to(device)

print('##### GENERATOR #####')
print(generator)
print('######################')

print('\n##### DISCRIMINATOR #####')
print(discriminator)
print('######################')

# optimizers
optim_g = optim.Adam(generator.parameters(), lr=learning_rate_g)
optim_d = optim.Adam(discriminator.parameters(), lr=learning_rate_d)

# loss function
criterion = nn.BCELoss()

losses_g = [] # to store generator loss after each epoch
losses_d = [] # to store discriminator loss after each epoch
images = [] # to store images generatd by the generator

# to create real labels (1s)
def label_real(size):
    data = torch.ones(size, 1)
    return data.to(device)

# to create fake labels (0s)
def label_fake(size):
    data = torch.zeros(size, 1)
    return data.to(device)

# function to create the noise vector
def create_noise(sample_size, nz):
    return torch.randn(sample_size, nz).to(device)

# to save the images generated by the generator
def save_generator_image(image, path):
    save_image(image, path)

# function to train the discriminator network
def train_discriminator(optimizer, data_real, data_fake):
    b_size = data_real.size(0)
    real_label = label_real(b_size)
    fake_label = label_fake(b_size)

    optimizer.zero_grad()

    output_real = discriminator(data_real)
    loss_real = criterion(output_real, real_label)

    output_fake = discriminator(data_fake)
    loss_fake = criterion(output_fake, fake_label)


    loss_real.backward()
    loss_fake.backward()
    optimizer.step()

    return loss_real + loss_fake

# function to train the generator network
def train_generator(optimizer, data_fake):
    b_size = data_fake.size(0)
    real_label = label_real(b_size)

    optimizer.zero_grad()

    output = discriminator(data_fake)
    loss = criterion(output, real_label)

    loss.backward()
    optimizer.step()

    return loss    

# create the noise vector
noise = create_noise(sample_size, nz)

generator.train()
discriminator.train()

for epoch in range(epochs):
    loss_g = 0.0
    loss_d = 0.0
    for bi, data in tqdm(enumerate(train_loader), total=int(len(train_data)/train_loader.batch_size)):
        image, _ = data
        image = image.to(device)
        b_size = len(image)
        # run the discriminator for k number of steps
        for step in range(k):
            data_fake = generator(create_noise(b_size, nz)).detach()
            data_real = image
            # train the discriminator network
            loss_d += train_discriminator(optim_d, data_real, data_fake)
        data_fake = generator(create_noise(b_size, nz))
        # train the generator network
        loss_g += train_generator(optim_g, data_fake)

    # create the final fake image for the epoch
    generated_img = generator(noise).cpu().detach()
    # make the images as grid
    generated_img = make_grid(generated_img)
    # save the generated torch tensor models to disk
    save_generator_image(generated_img, f"outputs/gen_img{epoch}.jpg")
    images.append(generated_img)
    epoch_loss_g = loss_g / bi # total generator loss for the epoch
    epoch_loss_d = loss_d / bi # total discriminator loss for the epoch
    losses_g.append(epoch_loss_g)
    losses_d.append(epoch_loss_d)
    
    print(f"Epoch {epoch} of {epochs}")
    print(f"Generator loss: {epoch_loss_g:.8f}, Discriminator loss: {epoch_loss_d:.8f}")

print('DONE TRAINING')
torch.save(generator.state_dict(), 'outputs/generator.pth')

# save the generated images as GIF file
imgs = [np.array(to_pil_image(img)) for img in images]
imageio.mimsave('outputs/generator_images.gif', imgs)

# plot and save the generator and discriminator loss
plt.figure()
plt.plot(losses_g, label='Generator loss')
plt.plot(losses_d, label='Discriminator Loss')
plt.legend()
plt.savefig('outputs/loss.jpg')



gan_upgrade.py
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torch.optim as optim
import torchvision.datasets as datasets
import imageio
import numpy as np
import matplotlib
import os
import cv2

from torchvision.utils import make_grid, save_image
from torch.utils.data import DataLoader
from matplotlib import pyplot as plt
from tqdm import tqdm
from PIL import Image

matplotlib.style.use('ggplot')


###########################################

# learning parameters
batch_size = 256
learning_rate_g = 0.0001
learning_rate_d = 0.00005
epochs = 1200
count = ""

###########################################


directory = "bs" + str(batch_size) + "_lrg" + str(learning_rate_g) + "_lrd" + str(learning_rate_d) + "_ep"+str(epochs) + count

def createFolder(directory):
    try:
        if not os.path.exists(directory):
            os.makedirs(directory)
    except OSError:
        print ('Error: Creating directory. ' +  directory)
 
createFolder("results/" + directory)
createFolder("results_gray/" + directory)


sample_size = 64 # fixed sample size
nz = 128 # latent vector size
k = 1 # number of steps to apply to the discriminator
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

transform = transforms.Compose([
                                transforms.ToTensor(),
                                transforms.Normalize((0.5,),(0.5,)),
])

to_pil_image = transforms.ToPILImage()

train_data = datasets.ImageFolder(
    root='input/data/motion',
    transform=transform
)
train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True)

class Generator(nn.Module):
    def __init__(self, nz):
        super(Generator, self).__init__()
        self.nz = nz
        self.main = nn.Sequential(
            nn.Linear(self.nz, 256),
            nn.LeakyReLU(0.2),

            nn.Linear(256, 512),
            nn.LeakyReLU(0.2),

            nn.Linear(512, 1024),
            nn.LeakyReLU(0.2),

            nn.Linear(1024, 924),
            nn.Tanh(),
        )

    def forward(self, x):
        return self.main(x).view(-1, 3, 22, 14)

class Discriminator(nn.Module):
    def __init__(self):
        super(Discriminator, self).__init__()
        self.n_input = 924
        self.main = nn.Sequential(
            nn.Linear(self.n_input, 1024),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),

            nn.Linear(1024, 512),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),

            nn.Linear(512, 256),
            nn.LeakyReLU(0.2),
            nn.Dropout(0.3),

            nn.Linear(256, 1),
            nn.Sigmoid(),
        )

    def forward(self, x):
        x = x.view(-1, 924)
        return self.main(x)

generator = Generator(nz).to(device)
discriminator = Discriminator().to(device)

print('##### GENERATOR #####')
print(generator)
print('######################')

print('\n##### DISCRIMINATOR #####')
print(discriminator)
print('######################')

# optimizers
optim_g = optim.Adam(generator.parameters(), lr=learning_rate_g)
optim_d = optim.Adam(discriminator.parameters(), lr=learning_rate_d)

# loss function
criterion = nn.BCELoss()

losses_g = [] # to store generator loss after each epoch
losses_d = [] # to store discriminator loss after each epoch
images = [] # to store images generatd by the generator

# to create real labels (1s)
def label_real(size):
    data = torch.ones(size, 1)
    return data.to(device)

# to create fake labels (0s)
def label_fake(size):
    data = torch.zeros(size, 1)
    return data.to(device)

# function to create the noise vector
def create_noise(sample_size, nz):
    return torch.randn(sample_size, nz).to(device)

# to save the images generated by the generator
def save_generator_image(image, path):
    save_image(image, path)

# function to train the discriminator network
def train_discriminator(optimizer, data_real, data_fake):
    b_size = data_real.size(0)
    real_label = label_real(b_size)
    fake_label = label_fake(b_size)

    optimizer.zero_grad()

    output_real = discriminator(data_real)
    loss_real = criterion(output_real, real_label)

    output_fake = discriminator(data_fake)
    loss_fake = criterion(output_fake, fake_label)


    loss_real.backward()
    loss_fake.backward()
    optimizer.step()

    return loss_real + loss_fake

# function to train the generator network
def train_generator(optimizer, data_fake):
    b_size = data_fake.size(0)
    real_label = label_real(b_size)

    optimizer.zero_grad()

    output = discriminator(data_fake)
    loss = criterion(output, real_label)

    loss.backward()
    optimizer.step()

    return loss    

# create the noise vector
noise = create_noise(sample_size, nz)

generator.train()
discriminator.train()

for epoch in range(epochs):
    loss_g = 0.0
    loss_d = 0.0
    for bi, data in tqdm(enumerate(train_loader), total=int(len(train_data)/train_loader.batch_size)):
        image, _ = data
        image = image.to(device)
        b_size = len(image)
        # run the discriminator for k number of steps
        for step in range(k):
            data_fake = generator(create_noise(b_size, nz)).detach()
            data_real = image
            # train the discriminator network
            loss_d += train_discriminator(optim_d, data_real, data_fake)
        data_fake = generator(create_noise(b_size, nz))
        # train the generator network
        loss_g += train_generator(optim_g, data_fake)

    # create the final fake image for the epoch
    generated_img = generator(noise).cpu().detach()
    # make the images as grid
    generated_img = make_grid(generated_img)
    # save the generated torch tensor models to disk
    save_generator_image(generated_img, "results/" + directory + f"/gen_img{epoch}.jpg")
    save_generator_image(generated_img, "results_gray/" + directory + f"/gen_img{epoch}.jpg")
    images.append(generated_img)
    epoch_loss_g = loss_g / bi # total generator loss for the epoch
    epoch_loss_d = loss_d / bi # total discriminator loss for the epoch
    losses_g.append(epoch_loss_g)
    losses_d.append(epoch_loss_d)
    
    print(f"Epoch {epoch} of {epochs}")
    print(f"Generator loss: {epoch_loss_g:.8f}, Discriminator loss: {epoch_loss_d:.8f}")

print('DONE TRAINING')

torch.save(generator.state_dict(), "results/" + directory + '/generator.pth')

# save the generated images as GIF file
imgs = [np.array(to_pil_image(img)) for img in images]
imageio.mimsave("results/" + directory + '/generator_images.gif', imgs)

# plot and save the generator and discriminator loss
plt.figure()
plt.plot(losses_g, label='Generator loss')
plt.plot(losses_d, label='Discriminator Loss')
plt.legend()
plt.savefig("losses/" + directory + ".jpg")


path = 'results_gray\\' + directory
imagePaths = [os.path.join(path,file_name) for file_name in os.listdir(path)]

for imagePath in imagePaths:
    img = Image.open(imagePath).convert('L')
    img_numpy = np.array(img, 'uint8')
    cv2.imwrite("results_gray\\" + directory + "\\" + imagePath.split("\\")[-1], img_numpy)

print("DONE Converting To Gray")








